### ISSUE TITLE
ESMX Build Architecture - Link to Component

### ISSUE THREAD

**Issue by danrosen25** [esmf-org/esmf#100]
I'm creating a nuopc-app-prototype that uses only CMake and I noticed the `<model>.cmake` files are manually [generated in the Makefile](https://github.com/esmf-org/nuopc-app-prototypes/blob/develop/ESMX_AtmOcnProto/Lumo/Makefile#L62) with an issue, they are defining the esmx_driver library linking.

The [install(EXPORT...) ](https://cmake.org/cmake/help/latest/command/install.html#export) function built into CMake will generate a `<model>.cmake` file for you BUT these files aren't intended to define linking from esmx_driver to the library because they don't know about the esmx_driver. We should modify the ESMX build system so that it can use the `<model>.cmake` files generated by CMake. If we do so then end-users won't even have to know about esmx_driver.

The[ esmxGen.py script](https://github.com/esmf-org/esmf/blob/develop/src/addon/ESMX/Driver/esmxGen.py) should really add the target_link_libraries(esmx_driver PUBLIC <model>)

We should also make an executable (or script) like ESMX_Build that calls
cmake -S$(ESMF_ESMXDIR) -B<SOME_BUILD_DIR>

**Comment by danrosen25** [comment]
@theurich Let's review this when you're back from your holiday weekend.

**Comment by theurich** [comment]
@danrosen25

I completely agree, ESMX should be using `<model>.cmake` exactly the way it is generated by CMake `install(EXPORT...)` function. Do you have a nuopc-app-prototype branch with the new/changed prototype where the model uses CMake that I could just use to see hoe the `<model>.cmake` looks like in that case? What you say about `target_link_libraries(esmx_driver PUBLIC )` makes perfect sense, but it would be nice to work on it within an actual prototype case.

Basically I am thinking if you have a CMake model case you can push, I would like to take it from here, make the ESMX changes, and also adjust the exiting two ESMX prototype examples (that use GNU Make) to match the standard  `<model>.cmake` format. Plus I will update any documentation to match all of this. I would hope to get that in quickly, today/tomorrow.

I also agree with you point about offering an convenience tool that calls `cmake -S$(ESMF_ESMXDIR) -B<SOME_BUILD_DIR>`. I have a similar note from last time I worked on ESMX before the release. Just ran out of time to get this in for 8.4.0. It is a good time now to bring this in.

**Comment by danrosen25** [comment]
@theurich 
I do have one and I'll push it. Do you have time today if I schedule a meeting and show you a few things and ideas that I have?

- Dan

**Comment by theurich** [comment]
@ufuk, good point about real world cases being even more complex. For NEPTUNE, where I just use the GNU Make approach to generate the `<model>.cmake` file, I also had to put in a lot more information to find all the many 3rd party dependencies during linking. I agree that we need to figure out how to do this properly through CMake `install(EXPORT...)`. I must assume that there is a way... but I also can see that the resulting `<model>.cmake` looks different than what we assume in ESMX right now.
Let's start with the simple prototype that Dan is providing, make it work, and then widen the scope to look at your CDEPS and other cases that use CMake to generate the `<model>.cmake`, which are more complex.

**Comment by uturuncoglu** [comment]
@theurich Sure. If we could have a generic solution that could be adapted by the user easily in their CMakeLists.txt file that would boost the usage of ESMX across the other modeling systems.

**Comment by danrosen25** [comment]
Here's the example I mentioned. Notice line 22 of build.sh, where I add target_link_libraries(esmx_driver PUBLIC mymodelphy) as a temporary fix. If you comment out this line then the `<model>.cmake` generated during the MyModel install step doesn't link esmx_driver to mymodelphy.

[https://github.com/esmf-org/nuopc-app-prototypes/tree/feature/esmx_cmake/ESMX_SingleModelProto](https://github.com/esmf-org/nuopc-app-prototypes/tree/feature/esmx_cmake/ESMX_SingleModelProto)

**Comment by theurich** [comment]
@danrosen25 
Very good. This looks like exactly what I need to work on it. At this point I think it will be tomorrow for me to get to it. It all looks pretty clear though from just quickly looking at it.

**Comment by danrosen25** [comment]
@theurich 
Great! Since you're busy today then let's meet tomorrow.

I have more ideas to discuss: We should build an application (or script) ESMX_Builder that can be added to the user/system path during ESMF installation.  This script should replace the build.sh. The script arguments can read in a build configuration file (esmxBuild.yaml) and/or other command line arguments. It can either build just the ESMX application or build all the components then the ESMX application or maybe even checkout components too.

We could potentially replace entire build systems with:
`> ESMX_Builder ufs_short_range_weather.yaml <options like --build-type=Debug --prefix=installation_location>`
this command on any system with ESMF installed in any directory and ending up with an executable
